### PR TITLE
[WIP] ENH: add Pyarrow csv engine

### LINKED ISSUE #23697: ENH: allow engine='pyarrow' in read_csv
we could conditionally use the new pyarrow csv parser as an engine (requires 0.11 IIRC). eventually leading to a replacement path for the existing code. There might be a number of restrictions on what options we can pass as the current parser is more full-featured, but I suspect most of the basic options work.

cc @gfyoung @pitrou @wesm @TomAugspurger

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #23697
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Picking up #31817. 

I'm putting this up to use CI for debugging. There are a bunch of comments in #31817 that I haven't addressed yet (working on them) - but I'll take feedback at any time

### CODE REVIEW

File: pandas/tests/io/parser/test_common.py [hunk 1]
```diff
@@ -64,13 +66,14 @@ def _set_noconvert_columns(self):
...
     }
+    parser.engine = "c"
     parser._engine = MyCParserWrapper(StringIO(data), **parser.options)
 
     result = parser.read()
     tm.assert_frame_equal(result, expected)
 
 
-def test_empty_decimal_marker(all_parsers):
+def test_empty_decimal_marker(all_parsers, pyarrow_xfail):
```
Comment (WillAyd) [inline pandas/tests/io/parser/test_common.py:76]: Can `pyarrow_xfail` be used as a decorator like `@skip_pyarrow` instead of supplied as an argument?

File: pandas/tests/io/parser/test_common.py [hunk 2]
```diff
@@ -83,6 +86,7 @@ def test_empty_decimal_marker(all_parsers):
         parser.read_csv(StringIO(data), decimal="")
 
 
+@skip_pyarrow
```
Comment (WillAyd) [inline pandas/tests/io/parser/test_common.py:89]: What percentage of tests in this module are we skipping? It might make more sense to reorganize into classes and apply this skip to the class rather than marking a ton of tests individually

File: pandas/tests/io/parser/test_common.py (hunk 1 above)
Comment (arw2019) [reply pandas/tests/io/parser/test_common.py:76]: yes. Done

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (arw2019) [reply pandas/tests/io/parser/test_common.py:90]: It's a lot (15% of the total) so it may be worth doing. Will look

File: pandas/tests/io/parser/test_parse_dates.py [hunk 3]
```diff
@@ -34,6 +34,8 @@
 else:
     date_strategy = st.datetimes()
 
+pytestmark = pytest.mark.usefixtures("pyarrow_skip")
```
Comment (lithomas1) [inline pandas/tests/io/parser/test_parse_dates.py:37]: These tests used to pass but are now broken since you removed np.asarray from the datetime handling. Please revert that change and changes to these tests. ( see 
https://github.com/pandas-dev/pandas/pull/31817#discussion_r513163314, for more context.)

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (lithomas1) [reply pandas/tests/io/parser/test_common.py:90]: @arw2019 if possible can you xfail instead of skip, that way if the tests somehow pass in the future, we can find and un xfail them.

File: pandas/tests/io/parser/test_skiprows.py [hunk 4]
```diff
@@ -14,6 +14,8 @@
 from pandas import DataFrame, Index
 import pandas._testing as tm
 
+pytestmark = pytest.mark.usefixtures("pyarrow_skip")
```
Comment (lithomas1) [inline pandas/tests/io/parser/test_skiprows.py:17]: @arw2019, I remember some of these tests used to pass while I was still working on the arrow csv reader. Can you make sure nothing broke during your refactors? Thanks.

File: pandas/tests/io/parser/test_common.py [hunk 5]
```diff
@@ -902,6 +916,7 @@ def test_read_csv_parse_simple_list(all_parsers):
     tm.assert_frame_equal(result, expected)
 
 
+@skip_pyarrow
 @tm.network
 def test_url(all_parsers, csv_dir_path):
```
Comment (lithomas1) [inline pandas/tests/io/parser/test_common.py:921]: FYI, this is a good test to check if code changes break the entire thing.

File: pandas/tests/io/parser/test_parse_dates.py (hunk 3 above)
Comment (arw2019) [reply pandas/tests/io/parser/test_parse_dates.py:37]: No argument that we need the `np.asarray` to make the datetime handling pass but I'm going to do that in a separate PR once this is in. This PR is intrinsically large, I want to keep the scope minimal

File: pandas/tests/io/parser/test_skiprows.py (hunk 4 above)
Comment (arw2019) [reply pandas/tests/io/parser/test_skiprows.py:17]: I skipped the tests to try to diagnose the timeouts on CI jobs. Agreed it's better to xfail if possible but currently multiple CI jobs are failing so aiming to get to green then look at re-activating these

**Comment by arw2019** [comment]
CI green.

**Comment by lithomas1** [comment]
@arw2019 Thanks for picking up where I left off. Can you go and rerun the benchmarks to ensure that no performance regressions occurred in the refactor, when this is ready? (Note: speedup should be based on number of cores)

File: doc/source/user_guide/io.rst [hunk 6]
```diff
@@ -158,9 +158,11 @@ dtype : Type name or dict of column -> type, default ``None``
   (unsupported with ``engine='python'``). Use ``str`` or ``object`` together
   with suitable ``na_values`` settings to preserve and
   not interpret dtype.
-engine : {``'c'``, ``'python'``}
-  Parser engine to use. The C engine is faster while the Python engine is
-  currently more feature-complete.
+engine : {``'c'``, ``'pyarrow'``, ``'python'``}
```
Comment (jreback) [inline doc/source/user_guide/io.rst:161]: versionadded 1.3

File: doc/source/user_guide/io.rst [hunk 7]
```diff
@@ -1602,11 +1604,17 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
 Specifying the parser engine
 ''''''''''''''''''''''''''''
 
-Under the hood pandas uses a fast and efficient parser implemented in C as well
-as a Python implementation which is currently more feature-complete. Where
-possible pandas uses the C parser (specified as ``engine='c'``), but may fall
-back to Python if C-unsupported options are specified. Currently, C-unsupported
-options include:
+Pandas currently supports three engines, the C engine, the python engine, and an optional
```
Comment (jreback) [inline doc/source/user_guide/io.rst:1607]: versionadded 1.3

say pyarrow engine is faster *on some workloads*)

File: doc/source/user_guide/io.rst [hunk 8]
```diff
@@ -1602,11 +1604,17 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
...
+pyarrow engine. The pyarrow engine is fastest, followed by the C and Python engines. However,
+the pyarrow engine is much less robust than the C engine, and the C engine is less feature-rich
```
Comment (jreback) [inline doc/source/user_guide/io.rst:1609]: the c-python engine are quite close, make that more apparent here.

File: doc/source/user_guide/io.rst [hunk 9]
```diff
@@ -1602,11 +1604,17 @@ Specifying ``iterator=True`` will also return the ``TextFileReader`` object:
...
+than the Python engine.
+
+Where possible pandas uses the C parser (specified as ``engine='c'``), but it may fall
+back to Python if C-unsupported options are specified. If pyarrow unsupported options are
+specified while using ``engine='pyarrow'``, the parser will throw an error.
+(a full list of unsupported options is available at ``pandas.io.parsers._pyarrow_unsupported``).
+
+Currently, C-unsupported options include:
```
Comment (jreback) [inline doc/source/user_guide/io.rst:1617]: add a list of pyarrow unsupported options here as well (maybe make a csv table of all of the options)

File: doc/source/whatsnew/v1.3.0.rst [hunk 10]
```diff
@@ -13,6 +13,11 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+read_csv() now accepts pyarrow as an engine
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`pandas.read_csv` now accepts engine="pyarrow" as an argument, allowing for faster csv parsing on multicore machines
+with pyarrow installed. See the :doc:`I/O docs </user_guide/io>` for more info. (:issue:`23697`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:20]: blank line here

use double backticks around ```engine='pyarrow'```

File: pandas/io/parsers.py [hunk 11]
```diff
@@ -2165,6 +2231,129 @@ def _maybe_parse_dates(self, values, index, try_parse_dates=True):
         return values
 
 
+class BytesIOWrapper:
```
Comment (jreback) [inline pandas/io/parsers.py:2234]: this should be in pandas/io/common.py; though don't we already have something similar?

File: pandas/io/parsers.py [hunk 12]
```diff
@@ -2165,6 +2231,129 @@ def _maybe_parse_dates(self, values, index, try_parse_dates=True):
...
+            in ("include_columns", "null_values", "true_values", "false_values")
+        }
+        self.read_options = {
+            "autogenerate_column_names": self.header is None,
+            "skip_rows": self.header
+            if self.header is not None
+            else self.kwds["skiprows"],
+        }
+
+    def _finalize_output(self, frame):
```
Comment (jreback) [inline pandas/io/parsers.py:2320]: it would be good (ideally as a pre-cursor) to make _finalize_output a method on ParserBase and implement it for c/python (instead of some of the code on L1120).

File: pandas/tests/io/parser/conftest.py [hunk 13]
```diff
@@ -63,21 +68,27 @@ def csv1(datapath):
...
+_all_parser_ids = [*_c_parser_ids, *_py_parser_ids, *_pyarrow_parser_ids]
 
 
 @pytest.fixture(params=_all_parsers, ids=_all_parser_ids)
 def all_parsers(request):
     """
     Fixture all of the CSV parsers.
     """
+    if request.param.engine == "pyarrow":
+        pytest.importorskip("pyarrow", "0.15.0")
```
Comment (jreback) [inline pandas/tests/io/parser/conftest.py:91]: is this min version mentioned anywhere in doc-strings?

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/io/parser/test_common.py:90]: its totally fine for now to xfail these, just as long as we have a consistent name)

File: pandas/tests/io/parser/test_common.py [hunk 14]
```diff
@@ -1260,16 +1307,17 @@ def test_float_parser(all_parsers):
     tm.assert_frame_equal(result, expected)
 
 
-def test_scientific_no_exponent(all_parsers_all_precisions):
-    # see gh-12215
-    df = DataFrame.from_dict({"w": ["2e"], "x": ["3E"], "y": ["42e"], "z": ["632E"]})
-    data = df.to_csv(index=False)
-    parser, precision = all_parsers_all_precisions
+# def test_scientific_no_exponent(all_parsers_all_precisions):
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:1310]: dont' comment code

File: pandas/tests/io/parser/test_common.py [hunk 15]
```diff
@@ -1351,35 +1402,37 @@ def test_numeric_range_too_wide(all_parsers, exp_data):
...
-
-        value = np.inf if exp > 0 else 0.0
-        expected = DataFrame({"data": [value]})
-    else:
-        expected = DataFrame({"data": [f"10E{exp}"]})
-
-    tm.assert_frame_equal(result, expected)
-
-
+# @skip_pyarrow
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:1405]: dont' comment code

File: pandas/tests/io/parser/test_dtypes.py [hunk 16]
```diff
@@ -16,590 +16,587 @@
...
-        np.random.rand(5, 2).round(4),
-        columns=list("AB"),
-        index=["1A", "1B", "1C", "1D", "1E"],
-    )
-
-    with tm.ensure_clean("__passing_str_as_dtype__.csv") as path:
-        df.to_csv(path)
-
-        result = parser.read_csv(path, dtype=dtype, index_col=0)
+@skip_pyarrow
```
Comment (jreback) [inline pandas/tests/io/parser/test_dtypes.py:22]: do this in a separate PR (I think you already are)

**Review by jreback — CHANGES_REQUESTED** [review]
some comments, generally looks ok. mainly

- precursor to try to unify finalize_output for c/python engines to make this a method on PyArrowEngine
- testing updates to make skipping 'easier' e.g. per module is best (i know you have something outstanding for that).

but generally +1 on merging after the above

thanks @lithomas1 and @arw2019 for picking this up

File: pandas/io/parsers.py (hunk 11 above)
Comment (arw2019) [reply pandas/io/parsers.py:2234]: xref https://github.com/pandas-dev/pandas/pull/37639#issuecomment-742278895 I checked with @twoertwein and we don't

Planning to move this to pandas/io/common.py and integrate it into `pandas.io.common.get_handle` to be consistent with what we do for other IO methods. 

xref https://github.com/pandas-dev/pandas/pull/37639#issuecomment-742650316 Might spin it off into a precursor PR since `BytesIOWrapper` is in principle useful outside the pyarrow reader

File: pandas/io/parsers.py (hunk 12 above)
Comment (arw2019) [reply pandas/io/parsers.py:2320]: Will do

File: pandas/tests/io/parser/conftest.py (hunk 13 above)
Comment (arw2019) [reply pandas/tests/io/parser/conftest.py:91]: No and this should just be removed

When this PR started pyarrow required version for this was higher than the repo-wide minimum. We have since bumped the minimum pyarrow version so no need for special handling here

Comment (lithomas1) [reply pandas/tests/io/parser/conftest.py:91]: @jreback @arw2019 
The min version for pandas in general is 0.15.0, and this code was intended to skip pyarrow tests for machines that don't have the required version of pyarrow installed.

Comment (jreback) [reply pandas/tests/io/parser/conftest.py:91]: to be honest for 1.3 we should just bump to 1.0 anyhow to avoid all of these complexites. let me make an issue.

Comment (arw2019) [reply pandas/tests/io/parser/conftest.py:91]: > @jreback @arw2019
> The min version for pandas in general is 0.15.0, and this code was intended to skip pyarrow tests for machines that don't have the required version of pyarrow installed.

Right in which case we want to use the global min version here instead of hard coding for whenever we decide to bump the global min

Comment (lithomas1) [reply pandas/tests/io/parser/conftest.py:91]: Yeah, no disagreements there, we can probably pull this from the dependencies version dict. However, without this, pyarrow tests will fail on machines where it isn't installed I think.

File: pandas/io/parsers.py [hunk 17]
```diff
@@ -2165,6 +2231,129 @@ def _maybe_parse_dates(self, values, index, try_parse_dates=True):
...
+                if is_integer(item):
+                    self.index_col[i] = frame.columns[item]
+            frame.set_index(self.index_col, drop=True, inplace=True)
+
+        if self.kwds.get("dtype") is not None:
+            frame = frame.astype(self.kwds.get("dtype"))
+        return frame
+
+    def read(self):
+        pyarrow = import_optional_dependency("pyarrow.csv")
```
Comment (lithomas1) [inline pandas/io/parsers.py:2342]: I am fairly certain that version checking is broken for submodule imports like this, which is part of the reason I was modifying import_optional_dependency. Can you make sure that this code errors with pyarrow < 0.15?

File: pandas/tests/io/parser/test_skiprows.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/tests/io/parser/test_skiprows.py:17]: I recall spending some time trying to get the full test suite to pass for the original PR. IIRC tests that passed in isolation would hang when run as part of the full test run.

so while ideally the tests should be xfailed, unless we can isolate the cause of the lockups, then skips may unfortunately be necessary.

File: pandas/io/parsers.py (hunk 17 above)
Comment (arw2019) [reply pandas/io/parsers.py:2342]: I can (but also that sounds like something that should have its own issue and a separate PR)

Comment (lithomas1) [reply pandas/io/parsers.py:2342]: Ok, I'll fix this as a followup

Comment (arw2019) [reply pandas/io/parsers.py:2342]: I pulled your fix from the old branch here and it works
https://github.com/pandas-dev/pandas/compare/master...arw2019:submodule-import-optional-dep?expand=1
the problem is the test (I haven't quite figured out how to add a mock submodule I think)

I'd say that's orthogonal to this PR so can do whenever

Comment (jreback) [reply pandas/io/parsers.py:2342]: yeah just make an issue for followups with checkboxes is fine

Comment (jreback) [reply pandas/io/parsers.py:2342]: different than #38872 as that is implementing features, these are more folllowups about style / testing and so on

File: doc/source/user_guide/io.rst (hunk 9 above)
Comment (arw2019) [reply doc/source/user_guide/io.rst:1617]: xref https://github.com/pandas-dev/pandas/issues/38863

File: pandas/io/parsers.py (hunk 17 above)
Comment (arw2019) [reply pandas/io/parsers.py:2342]: > yeah just make an issue for followups with checkboxes is fine

#38889

Comment (arw2019) [reply pandas/io/parsers.py:2342]: > Ok, I'll fix this as a followup

tracking at #38888

Comment (arw2019) [reply pandas/io/parsers.py:2342]: Will be fixed by #38925

**Comment by jreback** [comment]
@arw2019 i think the precursors were all merged, would be nice to get this in.

**Comment by arw2019** [comment]
ok - will rebase (sometime in the next few days - it's a little bit of work now) and ping when ready

**Comment by lithomas1** [comment]
Just a heads up that I'm going to try to revive this PR sometime soon-ish.

**Comment by simonjayhawkins** [comment]
> Just a heads up that I'm going to try to revive this PR sometime soon-ish.

OK. closing this one as stale.

**Comment by lithomas1** [comment]
Working on this rn, currently rebasing.
